src/MiddlewareFactoryInterface.php

```diff
@@ -0,0 +1,86 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Mezzio;
+
+use Laminas\Stratigility\Middleware\CallableMiddlewareDecorator;
+use Laminas\Stratigility\Middleware\RequestHandlerMiddleware;
+use Laminas\Stratigility\MiddlewarePipe;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+
+/**
+ * Marshal middleware for use in the application.
+ *
+ * This class provides a number of methods for preparing and returning
+ * middleware for use within an application.
+ *
+ * If any middleware provided is already a MiddlewareInterface, it can be used
+ * verbatim or decorated as-is. Other middleware types acceptable are:
+ *
+ * - PSR-15 RequestHandlerInterface instances; these will be decorated as
+ *   RequestHandlerMiddleware instances.
+ * - string service names resolving to middleware
+ * - arrays of service names and/or MiddlewareInterface instances
+ * - PHP callables that follow the PSR-15 signature
+ *
+ * Additionally, the class provides the following decorator/utility methods:
+ *
+ * - callable() will decorate the callable middleware passed to it using
+ *   CallableMiddlewareDecorator.
+ * - handler() will decorate the request handler passed to it using
+ *   RequestHandlerMiddleware.
+ * - lazy() will decorate the string service name passed to it, along with the
+ *   factory instance, as a LazyLoadingMiddleware instance.
+ * - pipeline() will create a MiddlewarePipe instance from the array of
+ *   middleware passed to it, after passing each first to prepare().
+ *
+ * @psalm-type InterfaceType = RequestHandlerInterface|RequestHandlerMiddleware|MiddlewareInterface
+ * @psalm-type CallableType = callable(ServerRequestInterface, RequestHandlerInterface): ResponseInterface
+ * @psalm-type MiddlewareParam = string|InterfaceType|CallableType|list<string|InterfaceType|CallableType>
+ */
+interface MiddlewareFactoryInterface
```

Comment:
I'd prefer having `MiddlewareInterface` as a return value for all methods which do return a middleware. We can keep the more-specific return value in the `MiddlewareFactory` for BC reasons but I haven't found any usage where it is necessary to have the exact implementing class provided as the return value.

Reply:
The `prepare` method in both `MiddlewareFactory` and `MiddlewareFactoryInterface` already returns `MiddlewareInterface`.
Not sure I follow where you mean that we should use `MiddlewareInterface` instead of `MiddlewareFactoryInterface`.